### DISCUSSION TITLE
cmake error emgucv 4.5.4 [ flag BUILD_opencv_world ]

### DISCUSSION THREAD

**Discussion by md3848** [emgucv/emgucv#640] [Q&A] [answered]
**When I enabled this flag [ BUILD_opencv_world ], cmake showed this error** :

CMake Error at opencv_contrib/modules/xobjdetect/CMakeLists.txt:4 (add_subdirectory):
  add_subdirectory not given a binary directory but the given source
  directory
  "F:/Windows/Programing/OpenCV/S.W/EmguCV/4.5.4/emgucv/opencv_contrib/modules/xobjdetect/tools"
  is not a subdirectory of
  "F:/Windows/Programing/OpenCV/S.W/EmguCV/4.5.4/emgucv/opencv/modules/world".
  When specifying an out-of-tree source a binary directory must be explicitly
  specified.
Call Stack (most recent call first):
  opencv/modules/world/CMakeLists.txt:13 (include)
  opencv/modules/world/CMakeLists.txt:32 (include_one_module)

**Comment by emgucv** [comment, marked as answer]
We have not tested building opencv_world. The module is not used in Emgu CV. 

The "cvextern" project replaces "opencv_world" in our implementation.

**Comment by md3848** [comment]
In opencv we have a flag [BUILD_opencv_world] that when we activate it converts all dlls to 1 dll
I thought in emgucv this flag [opencv_world] does the same thing ...
thanks for your response.

**Comment by emgucv** [comment]
On windows, the "cvextern.dll" files contains all opencv modules in a single dll, there is not need to turn on opencv_world when building the native binary for Emgu CV.

**Comment by md3848** [comment]
but why my "cvextern.dll" file size is 10mb! [ I built emgucv with cmake ]

![image](https://user-images.githubusercontent.com/17591943/140898482-02e961b3-141d-4788-8e05-0e1cc7bfd8b3.png)

And when I use these dlls, Visual Studio shows the following error :



The type initializer for 'Emgu.CV.CvInvoke' threw an exception.

System.TypeInitializationException: The type initializer for 'Emgu.CV.CvInvoke' threw an exception.
 ---> System.DllNotFoundException: Unable to load DLL 'cvextern' or one of its dependencies: The specified module could not be found. (0x8007007E)
   at Emgu.CV.CvInvoke.RedirectError(CvErrorCallback errorHandler, IntPtr userdata, IntPtr prevUserdata)
   at Emgu.CV.CvInvoke..cctor()
   --- End of inner exception stack trace ---
   at Emgu.CV.Image`2..ctor(String fileName)
   at WpfApp10.MainWindow.Button_OpenInputImage_Click(Object sender, RoutedEventArgs e) in C:\Users\mahdi\source\repos\WpfApp10\WpfApp10\MainWindow.xaml.cs:line 43

**Comment by emgucv** [comment]
The Emgu CV build script, follwing this instructions:
https://www.emgu.com/wiki/index.php/Download_And_Installation#Configuring_the_project
will build without cuda, all opencv modules will be build as static library and linked into a single "cvextern.dll" file.

Your cmake configuration seems to have the opencv modules compiled as dynamic library. Please make sure they are built as static libraray and they will be static linked into a single "cvextern.dll".

When building with CUDA support, because we are not able to figure out how to static linking using NVCC, if CUDA is enabled, the opencv modules will be compiled as dynamic library, similar to your screen shot, in which case the "cvextern.dll" will not contains all binaries, and will have dependency on the additional "opencv_{module}.dll" files.

**Comment by md3848** [comment]
> @emgucv : When building with CUDA support
1) I do not use CUDA.

> @emgucv : Your cmake configuration seems to have the opencv modules compiled as dynamic library
2) How? Which flag are you talking about in cmake?

3) In the image below [red markers] I have to build them [with cmake and visual studio] and then copy them to the emgucv source code folder? Or just copy the source code to the emgucv folder [vtk, hdf5, harfbuzz, ...]
3.1) I am copying the source code of these files : opencv, opencv_contrib, opencv_extra
3.2) i build [cmake + vs] then copy this files : vtk
3.3) I do not use these folders [cause errors] : harfbuzz, hdf5
![image](https://user-images.githubusercontent.com/17591943/141072968-72a1ef56-46dc-4f06-946a-171c0e690cb8.png)

4) Should I copy folders [vtk, hdf5, harfbuzz, ...] then run this "Build_Binary_x86-64_doc.bat" file first, then build with cmake, then build with Visual Studio? I'm right?
